tests/TestCase/Database/ConnectionTest.php

```diff
@@ -987,4 +988,116 @@ public function testSchemaCollection()
+                return true;
+            });
+
+            return true;
+        });
+        $this->assertFalse($this->connection->inTransaction());
+
+        $this->connection->transactional(function () {
+            $this->connection->transactional(function () {
+                return true;
+            });
+
+            return false;
+        });
+        $this->assertFalse($this->connection->inTransaction());
+
+        $this->connection->transactional(function () {
+            $this->connection->transactional(function () {
+                return false;
+            });
+
+            return false;
+        });
+        $this->assertFalse($this->connection->inTransaction());
+    }
+
+    /**
+     * Tests that not allowed nesting of commit/rollback operations throws
+     * a NestedTransactionRollbackException.
+     *
+     * @return void
+     * @expectedException \Cake\Database\Exception\NestedTransactionRollbackException
+     * @expectedExceptionMessage Cannot commit transaction - rollback() has been already called in the nested transaction
+     */
+    public function testNestedTransactionRollbackExceptionThrown()
+    {
+        // ROLLBACK -> COMMIT
+        $this->connection->transactional(function () {
+            $this->connection->transactional(function () {
+                return false;
+            });
+
+            return true;
+        });
+    }
+
+    /**
+     * Tests mor detail about that not allowed nesting of rollback/commit
+     * operations throws a NestedTransactionRollbackException by trace.
+     *
+     * @return void
+     */
+    public function testNestedTransactionRollbackExceptionByTrace()
+    {
+        $rollbackSourceLine = -1;
+        $nestedTransactionStates = [];
+
+        $e = null;
+        try {
+            $this->connection->transactional(function () use (&$rollbackSourceLine, &$nestedTransactionStates) {
+                $nestedTransactionStates[] = $this->connection->wasNestedTransactionRolledback();
+
+                $this->connection->transactional(function () {
+                    return true;
+                });
+
+                $this->connection->transactional(function () use (&$rollbackSourceLine, &$nestedTransactionStates) {
+                    $nestedTransactionStates[] = $this->connection->wasNestedTransactionRolledback();
+
+                    $this->connection->transactional(function () {
+                        return false;
+                    });
+                    $rollbackSourceLine = __LINE__ - 1;
+
+                    $nestedTransactionStates[] = $this->connection->wasNestedTransactionRolledback();
+
+                    return true;
+                });
+
+                $this->connection->transactional(function () {
+                    return false;
+                });
+
+                $nestedTransactionStates[] = $this->connection->wasNestedTransactionRolledback();
+
+                return true;
+            });
```

Comment:
Should you have `$this->fail(..)` here to ensure that the `NestedTransactionRollbackException` was thrown.